canonical/cloud-init
# Fix unpickle for source paths missing run_dir

## Proposed Commit Message
Fix unpickle for source paths missing run_dir

On the source class, we require the use of paths.run_dir to perform some operations. On older cloud-init version, the Paths class does not have the run_dir attribute. To fix that, we are now manually adding that attribute in the source paths
object if doesn't exist in the unpickle operation.

LP: #1899299

## Test Steps
To reproduce this issue, run the following script:
```sh
#!/bin/sh
set -x

series=trusty
name=$series-upgrade

lxc delete --force $name
lxc launch ubuntu-daily:$series $name
sleep 10

lxc exec $name -- add-apt-repository ppa:ua-client/stable -y
lxc exec $name -- sudo apt-get update
lxc exec $name -- sudo apt-get install ubuntu-advantage-tools -y
lxc exec $name -- ua version
lxc exec $name -- sudo ua attach <uaclient-token>
lxc exec $name -- sudo ua status
lxc exec $name -- sudo add-apt-repository ppa:cloud-init-dev/daily -y
lxc exec $name -- sudo apt-get update
lxc exec $name -- sudo mkdir -p /etc/update-manager/release-upgrades.d  
lxc exec $name -- sh -c "echo '[Sources]\nAllowThirdParty=yes' > allow.cfg"
lxc exec $name -- sudo mv allow.cfg /etc/update-manager/release-upgrades.d   
lxc exec $name -- sudo apt-get upgrade -y
lxc exec $name -- sudo apt-get dist-upgrade -y
lxc exec $name -- sudo do-release-upgrade --frontend DistUpgradeViewNonInteractive
lxc exec $name -- sudo ua status --wait

lxc exec $name -- sudo reboot
sleep 60
lxc exec $name -- sudo ua status --wait

set +x
```

By looking at the `/var/log/ubuntu-advantage.log` we will see that the error is because `cloud-id` fails with the error:
```
AttributeError: 'Paths' object has no attribute 'run_dir'
```

We can confirm that by running `cloud-id` on the container directly.

To verify that the fix works, run the following commands:
```
lxc file push cloudinit/helpers.py trusty-upgrade/usr/lib/python3/dist-packages/cloudinit/helpers.py
lxc exec trusty-upgrade -- reboot
sleep 30
lxc exec trusty-upgrade -- cloud-id
lxc exec trusty-upgrade -- ua status
```
## Checklist:
